### DISCUSSION TITLE
atomWithImmer not resolving the types hint properly

### DISCUSSION THREAD

**Discussion by khanakia** [pmndrs/jotai#2593] [Q&A]
## Summary

I have this code and when i mouse hover to `immerSetup.screen` in vscode it shows any

How do I make it typesafe ?

```ts
type Setup = {
  screen: string;
};

const defaultSetup = {
  screen: "step1",
};

// Original atom.
const setupImmerAtom = atomWithImmer<Setup>(defaultSetup);

const setScreen = (screen: Screen) => {
  const [setupAtom, setAtomSetup] = useAtom(setupImmerAtom);
  setAtomSetup((immerSetup) => {
    immerSetup.screen = screen;
    return immerSetup;
  });
};
```

![Screenshot 2024-06-01 at 7 07 33 PM](https://github.com/pmndrs/jotai/assets/2174170/1faf20c5-1550-49ce-85f9-3807d4a1da71)
![Screenshot 2024-06-01 at 7 18 47 PM](https://github.com/pmndrs/jotai/assets/2174170/7bf0337a-ba07-4ecc-87fc-86ac9106ff2f)

On debugging further, I found out this

```
// actual type definition /node_modules/jotai-immer/dist/atomWithImmer.d.ts
export declare function atomWithImmer<Value>(initialValue: Value): WritableAtom<Value, [Value | ((draft: Draft<Value>) => void)], void>;

// resolving as this
atomWithImmer<Setup>(initialValue: Setup): WritableAtom<Setup, [Setup | ((draft: Draft<Value>) => void)], void>

// IT should resolve as this `Draft<Value>` to `Draft<Setup>`
atomWithImmer<Setup>(initialValue: Setup): WritableAtom<Setup, [Setup | ((draft: Draft<Setup>) => void)], void>

```

By setting the generic type explicitly in `atomWithImmer<Setup>`, TypeScript should correctly infer `Draft<Setup>` instead of `Draft<Value>`.

## Link to reproduction
No need just a typescript hint error

## Check List

Please do not ask questions in issues.

- [x] I've already opened a [discussion](https://github.com/pmndrs/jotai/discussions) before opening this issue, or already discussed in other media.

Please include a minimal reproduction.

- [ ] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

**Comment by dai-shi** [comment]
Does this help? https://tsplay.dev/wjLgYm
